### PR TITLE
Hook abstraction lift

### PR DESCRIPTION
**Pull request by laserkelvin**
# ALCHEMI Toolkit Pull Request

## Description

This PR lifts the hook abstraction outside of being dynamics only, with the intention to make it generally usable across other (future) workflows/pipelines.

The main change to the core protocol itself is two-fold:

1. Adding a `HookContext` dataclass, which provides a mechanism for things like dynamics or training orchestrators to pass relevant information into the firing hook,
2. Making the `stage` argument to `Hook.__call__` be a generic `Enum`, and in doing so, making it the responsibility of the hook to implement a `_runs_on_stage` method that determines whether or not the hook should fire at a given stage. 

Consequently, this allows a hook to be defined and used in one or more workflows, without *needing* to modify its logic or duplicate it for another category (the neighbor list hook is one such example of this). Users can define a maximally general hook (by not specializing on the stage `Enum`), and/or can define workflow-specific behavior using `plum.dispatch`: the documentation has been updated to include these patterns.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Changes Made

- Defined namespace and moved protocol into `nvalchemi.hooks`
- Introduce a new `HookContext` dataclass that provides a storage mechanism for workflow orchestrators to pass context into firing hooks
- Introduce a `HookRegistryMixin`, which abstracts out hook management from workflow orchestrator classes; e.g. `BaseDynamics` now inherits from this as well, and I foresee others like training/active learning to do so
- Predefines a `TrainingStage` `Enum` and namespace, just to add the test coverage for another category
- Refactored existing hooks to adhere to the new protocol

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

## Additional Notes

A good sequence to follow:

- `nvalchemi/hooks/_protocol.py` gives the revised protocol basis for hooks
- `nvalchemi/hooks/_context.py` implements the `HookContext` dataclass
- `nvalchemi/hooks/_registry.py` implements the mixin used to manage hooks
- `nvalchemi.dynamics.base.BaseDynamics` shows the updated inheritance
- `hvalchemi/hooks/*.py` shows concrete implementations of hooks. `snapshot.py` is a straightforward implementation, and `neighbor_list.py` shows how to do a shareable hook

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemi/dynamics/demo.py [hunk 1]
```diff
@@ -82,7 +82,7 @@ class DemoDynamics(BaseDynamics):
         The integration timestep.
     step_count : int
         The current step number.
-    hooks : dict[HookStageEnum, list[Hook]]
+    hooks : dict[DynamicsStage, list[Hook]]
```
Comment (zubatyuk) [inline nvalchemi/dynamics/demo.py]: Should it be `list[Hook]`?

File: nvalchemi/hooks/_context.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    """
+
+    batch: Batch
+    step_count: int
+    model: BaseModelMixin | None = None
+    loss: torch.Tensor | None = None
+    optimizer: torch.optim.Optimizer | None = None
+    lr_scheduler: object | None = None
+    gradients: dict[str, torch.Tensor] | None = None
+    converged_mask: torch.Tensor | None = None
```
[the remark this answers is not in the document]
Comment (dallasfoster) [reply nvalchemi/hooks/_context.py:64]: @laserkelvin beyond this, I think we should consider whether we want to adopt indices vs. masks more broadly through the hooks. I would like to try and avoid syncs as much as possible. (Hence mask.any() is probably the least disruptive).

File: nvalchemi/dynamics/hooks/logging.py [hunk 3]
```diff
@@ -308,7 +335,8 @@ def _compute_columns(
 
         if self.custom_scalars:
             for name, fn in self.custom_scalars.items():
-                val = fn(batch, dynamics)
+                # custom_scalars receive (ctx,) — a HookContext
+                val = fn(ctx)
                 if isinstance(val, torch.Tensor):
                     td.set(name, val)
```
[the remark this answers is not in the document]
Comment (dallasfoster) [reply nvalchemi/dynamics/hooks/logging.py:341]: What is the desired behavior here with ctx?

File: nvalchemi/hooks/_context.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (laserkelvin) [reply nvalchemi/hooks/_context.py:64]: Addressed in 95e0e8c

File: nvalchemi/dynamics/hooks/logging.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (laserkelvin) [reply nvalchemi/dynamics/hooks/logging.py:341]: > What is the desired behavior here with ctx?

The optional aspect was left over from prior - it shouldn't be optional and fixed in 1e18131

File: nvalchemi/dynamics/demo.py (hunk 1 above)
Comment (laserkelvin) [reply nvalchemi/dynamics/demo.py]: Fixed in 140b812
